src/Grpc.AspNetCore.Server.ClientFactory/ContextPropagationInterceptor.cs

```diff
@@ -0,0 +1,172 @@
+                    call.ResponseStream,
+                    call.ResponseHeadersAsync,
+                    call.GetStatus,
+                    call.GetTrailers,
+                    () => { call.Dispose(); cts.Dispose(); });
+            }
+        }
+
+        public override AsyncServerStreamingCall<TResponse> AsyncServerStreamingCall<TRequest, TResponse>(TRequest request, ClientInterceptorContext<TRequest, TResponse> context, AsyncServerStreamingCallContinuation<TRequest, TResponse> continuation)
+        {
+            var call = continuation(request, ConfigureContext(context, out var cts));
+            if (cts == null)
+            {
+                return call;
+            }
+            else
+            {
+                return new AsyncServerStreamingCall<TResponse>(
+                    call.ResponseStream,
+                    call.ResponseHeadersAsync,
+                    call.GetStatus,
+                    call.GetTrailers,
+                    () => { call.Dispose(); cts.Dispose(); });
+            }
+        }
+
+        public override AsyncUnaryCall<TResponse> AsyncUnaryCall<TRequest, TResponse>(TRequest request, ClientInterceptorContext<TRequest, TResponse> context, AsyncUnaryCallContinuation<TRequest, TResponse> continuation)
+        {
+            var call = continuation(request, ConfigureContext(context, out var cts));
+            if (cts == null)
+            {
+                return call;
+            }
+            else
+            {
+                return new AsyncUnaryCall<TResponse>(
+                    call.ResponseAsync,
+                    call.ResponseHeadersAsync,
+                    call.GetStatus,
+                    call.GetTrailers,
+                    () => { call.Dispose(); cts.Dispose(); });
+            }
+        }
+
+        public override TResponse BlockingUnaryCall<TRequest, TResponse>(TRequest request, ClientInterceptorContext<TRequest, TResponse> context, BlockingUnaryCallContinuation<TRequest, TResponse> continuation)
+        {
+            var response = continuation(request, ConfigureContext(context, out var cts));
+            cts?.Dispose();
+            return response;
+        }
+
+        private ClientInterceptorContext<TRequest, TResponse> ConfigureContext<TRequest, TResponse>(ClientInterceptorContext<TRequest, TResponse> context, out CancellationTokenSource? linkedCts)
+            where TRequest : class
+            where TResponse : class
+        {
+            linkedCts = null;
+
+            var options = context.Options;
+            var serverCallContext = GetServerCallContext();
+
+            // Use propagated deadline if it is smaller than the specified deadline
+            if (serverCallContext.Deadline < context.Options.Deadline.GetValueOrDefault(DateTime.MaxValue))
+            {
+                options = options.WithDeadline(serverCallContext.Deadline);
+            }
+
+            if (serverCallContext.CancellationToken.CanBeCanceled)
```

Comment:
nit: the token will always be cancellable (as all calls can be cancelled by the client).

Reply:
On Kestrel someone could set the request CancellationToken to CancellationToken.None. Not likely, but a simple if check doesn't really have any downsides 😄